### ISSUE TITLE
Judder/frame pacing issue

### ISSUE THREAD

**Issue by BarbaxX15** [xbmc/xbmc#25681]
### Describe the bug
Here is a clear and concise description of what the problem is:

Up until the 30 minute mark videos run smoothly, then judder begins and gets worse over time.

- The playback device is an Android TV Dongle (Benq QS01) running Android TV 12.
- I have set frame rate matching to "ON", so 24p content plays at 24Hz (TV refresh rate switches).
- The source files are located on my NAS and I have tried playing them via Kodi directly (source = network location) and via the addon 
   "PlexKodiConnect", since I am running a Plex server as well.
- The judder also happens in the official Plex app on this device after 30 minutes.
- The judder does NOT happen on Amazon Prime Video [EDIT: Also happens in here when set to 24Hz!]
- The videos/movies play fine on other playback devices (Nvidia Shield, PC), in Kodi and the Plex app.

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

Video should play smoothly throughout.

## Actual Behavior

Video starts to judder after the 30 minute mark.

## Possible Fix

In combination with Android TV's dev menu option "Disable HW overlay", the only way I have found to mitigate the issue is enabling the Player Debug Info OSD in the top left corner. As long as this OSD is on, the video runs smoothly.

### To Reproduce

1. Start video playback from NAS/Plex
2. Play for 30 minutes

### Debuglog

Debug logs should not be necessary, since Kodi does not register the frame drops in the debug and player osd menus.
This issue seems to be connected to Android TV 12's handling of 24p content.

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

### Screenshots 
Here are some links or screenshots to help explain the problem:

I have uploaded a video of the issue here (you can see when debug OSD is on): [Kodi frame pacing issue](https://photos.onedrive.com/share/5CF03C33F12F19B5!20809?cid=5CF03C33F12F19B5&resId=5CF03C33F12F19B5!20809&authkey=!AH-2WTbZXTMWfkk&ithint=video&e=KOhNRw)

### Your Environment
Used Operating system:

 - [X] Android
 - [ ] iOS
 - [ ] tvOS
 - [ ] Linux
 - [ ] macOS
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name: Android TV 12
 - Kodi version: 21.1

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by thexai** [comment]
> The judder also happens in the official Plex app on this device after 30 minutes.

Then if is not only Kodi, not seems "kodi bug" is more a device hardware issue.

> The videos/movies play fine on other playback devices (Nvidia Shield, PC), in Kodi and the Plex app.

This points to same direction (not Kodi issue)

The fact that it happens after 30 minutes makes me think that the device is overheating and throttling the CPU so that it doesn't keep heating up. With the CPU limited then there are performance issues.

The reason why it doesn't happen while there is debug OSD may simply be that more CPU is used while the overlay is on. It may seem like a contradiction but if more CPU is required this probably forces a frequency one step higher. That is, the CPU performs better in this case while the OSD is on.

Solutions: 

- Improve the cooling of the device. Maybe it is defective: the thermal paste is bad or it doesn't make contact, etc.
- Maybe there is some background process or service that uses CPU unnecessarily. Check the background processes with adb.

Maybe related to https://github.com/xbmc/xbmc/issues/24980 and https://github.com/xbmc/xbmc/pull/24986#pullrequestreview-1999469591

You may try disable `com.google.android.tvrecommendations` service.

**Comment by BarbaxX15** [comment]
First of all thank you for the reply.

I concur that it is an issue with the hardware, but I hoped that the player debug info workaround could point me in the right direction for finding the root cause.

I can exclude thermal throttling as the device does not feel hotter than idle when the issue occurs. 
In addition to that, the issue occurs at _exactly_ the same time every time and quickly stopping playback (not pausing) and resuming from the last timestamp plays smoothly again for exactly that time period.

The device uses hardware decoding for all my videos (H264 and H265), so CPU load should be quite low actually. 
Does switching on the OSD switch the renderer to do software decoding?

What's weird is that the OSD only helps if the Android option "Disable HW overlay" is enabled...
I will try disabling the service you mentioned.

**Comment by thexai** [comment]
> What's weird is that the OSD only helps if the Android option "Disable HW overlay" is enabled...

Disable HW overlay helps to use more CPU (cpu sleeps less) and less energy saving profile. Enable Debug OSD in Kodi has similar effect because increases CPU usage. Video playback itself uses very little CPU because is HW accelerated (almost GPU usage only).

This seems a firmware bug that causes excess energy save when low CPU usage. Poor scheduling of power levels for different workloads (simplifying a lot).

It's not related to Kodi at all...

See also:

https://www.reddit.com/r/OnnStreamingTV/comments/1cl1w4p/video_starts_to_stutter_after_30_minutes_on_new/
https://www.reddit.com/r/AndroidTV/comments/1cl1u9j/video_starts_to_stutter_on_new_onn_pro_4k_after/

**Comment by BarbaxX15** [comment]
> It's not related to Kodi at all...
I never meant to state that this is an issue with Kodi. 
As I mentioned, the same issue occurs in other apps, but the fact that Kodi is the only media player that can mitigate the issue using the debug OSD, made me wonder how this overlay is handled and ask for further insight, which I want to thank you for.

I have done some further investigation based on your assumptions and just want to share them here in case someone else is struggling with this issue:

- ADB logs don't show a difference between the time the video runs smoothly and when the judder appears
- Using ADB's "top" (process monitor) I could see that Kodi uses 70% CPU when playing the video
- Temps are between 74-78°C
- Switching on the debug menu with "Disable HW Overlays" enabled spikes the CPU to 100-110%, which makes the video run smoothly again

So I guess you are right, that the CPU scheduling seems to enter a different power state after 30 minutes (despite CPU usage being constant) and messes up the rendering. Turning on the debug menu and increasing the CPU usage changes this power state and fixes the rendering, but once the CPU drops to the prior utilization after disabling the overlay, the previous power state becomes active again and the choppy playback returns.

Summing up, the only options I have are:

1. Downgrading the hardware (issue was not present before Android 12 update)
2. Playing content with 3:2 pulldown.
3. Stopping and resuming playback every 30 minutes.
4. Waiting for an OS/firmware update (unlikely, as device is discontinued)


Thank you for your support.

**Comment by mycek91** [comment]
Hi there,
I can confirm that exactly the same behaviour happens on Strong Leap-s1, choppy playback starts just after the 30 minute mark.
The problem does not occur when framerate switching is disabled in Kodi and framerate switching is handled by Tvquickactions.
And it only happens when autoframerate is enabled in Kodi and the video is 23.976 fps or 24 fps.
